src/Dict.cc

```diff
@@ -1146,4 +1458,168 @@ void Dictionary::StopIteration(IterCookie* cookie) const
 	dp->StopIterationNonConst(cookie);
 	}
 
+#pragma GCC diagnostic pop
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// New Iteration
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+DictIterator::DictIterator(const Dictionary* d, detail::DictEntry* begin, detail::DictEntry* end)
+	: curr(begin), end(end)
+	{
+	// Make sure that we're starting on a non-empty element.
+	while ( curr != end && curr->Empty() )
+		++curr;
+
+	// Cast away the constness so that the number of iterators can be modified in the dictionary. This does
+	// violate the constness guarantees of const-begin()/end() and cbegin()/cend(), but we're not modifying the
+	// actual data in the collection, just a counter in the wrapper of the collection.
+	dict = const_cast<Dictionary*>(d);
+	dict->num_iterators++;
+	}
+
+DictIterator::~DictIterator()
+	{
+	if ( dict->num_iterators > 0 )
+		dict->num_iterators--;
+	}
+
+DictIterator& DictIterator::operator++()
+	{
+	// The non-robust case is easy. Just advanced the current position forward until you find
+	// one isn't empty and isn't the end.
+	do {
+		++curr;
+		}
+	while ( curr != end && curr->Empty() );
+
+	return *this;
+	}
+
+
+
+
+std::shared_ptr<RobustDictIterator> Dictionary::MakeRobustIterator()
+	{
+	if ( ! iterators )
+		iterators = new std::vector<std::shared_ptr<RobustDictIterator>>;
+
+	num_iterators++;
+	auto iter = std::make_shared<RobustDictIterator>(this);
+	iterators->push_back(iter);
+
+	// Advance the iterator once so we're at the first position.
+	++(*iter);
```

Comment:
This doesn't work if I continue using `shared_ptr`/`shared_from_this` because it can't find the constructors when it tries to instantiate the templates during the build. I'll revisit it after I sort out your other comment about that stuff. That said, it does work fine for the non-robust iterator.